pkg/crds/upgrade.go

```diff
@@ -0,0 +1,148 @@
+package crds
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/k8ssandra/k8ssandra/pkg/helmutil"
+	"gopkg.in/yaml.v3"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	k8syaml "k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/clientcmd/api"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Upgrader is a utility to update the CRDs in a helm chart's pre-upgrade hook
+type Upgrader struct {
+	client    client.Client
+	namespace string
+}
+
+// NewWithClient returns a new Upgrader client using the given controller-runtime client.Client
+func NewWithClient(c client.Client) (*Upgrader, error) {
+	return &Upgrader{
+		client: c,
+	}, nil
+}
+
+// New returns a new Upgrader client
+func New(namespace string) (*Upgrader, error) {
+	_ = api.AddToScheme(scheme.Scheme)
+	c, err := client.New(ctrl.GetConfigOrDie(), client.Options{Scheme: scheme.Scheme})
+	if err != nil {
+		log.Fatal(err)
+		return nil, err
+	}
+
+	return &Upgrader{
+		client:    c,
+		namespace: namespace,
+	}, nil
+}
+
+// Upgrade installs the missing CRDs or updates them if they exists already
+func (u *Upgrader) Upgrade(targetVersion string) error {
```

Comment:
It would be nice to have some debug logging in here. I mentioned in my previous comment how I hard coded the value of `--targetversion`. I did the upgrade and as a quick test did `kubectl get crd cassandradatacenters.cassandra.datastax.com -o yaml | grep configSecret`. I did not get any results back. Then I uninstalled, deleted CRDs, and reinstalled with `helm install test k8ssandra/k8ssandra --version 1.2.0-20210514022645-da7547a5`. This time when I the preceding grep I get back results as expected.